apm-sniffer/apm-sdk-plugin/dubbo-3.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/asf/dubbo3/DubboInterceptor.java

```diff
@@ -0,0 +1,211 @@
+ * limitations under the License.
+ *
+ */
+
+package org.apache.skywalking.apm.plugin.asf.dubbo3;
+
+import org.apache.dubbo.common.constants.CommonConstants;
+import org.apache.dubbo.common.URL;
+import org.apache.dubbo.rpc.Invocation;
+import org.apache.dubbo.rpc.Invoker;
+import org.apache.dubbo.rpc.Result;
+import org.apache.dubbo.rpc.RpcContext;
+import org.apache.dubbo.rpc.RpcContextAttachment;
+import org.apache.dubbo.rpc.RpcException;
+import org.apache.dubbo.rpc.RpcServiceContext;
+import org.apache.skywalking.apm.agent.core.context.CarrierItem;
+import org.apache.skywalking.apm.agent.core.context.ContextCarrier;
+import org.apache.skywalking.apm.agent.core.context.ContextManager;
+import org.apache.skywalking.apm.agent.core.context.tag.Tags;
+import org.apache.skywalking.apm.agent.core.context.trace.AbstractSpan;
+import org.apache.skywalking.apm.agent.core.context.trace.SpanLayer;
+import org.apache.skywalking.apm.agent.core.plugin.interceptor.enhance.EnhancedInstance;
+import org.apache.skywalking.apm.agent.core.plugin.interceptor.enhance.InstanceMethodsAroundInterceptor;
+import org.apache.skywalking.apm.agent.core.plugin.interceptor.enhance.MethodInterceptResult;
+import org.apache.skywalking.apm.network.trace.component.ComponentsDefine;
+import org.apache.skywalking.apm.util.StringUtil;
+
+import java.lang.reflect.Method;
+
+/**
+ * {@link DubboInterceptor} define how to enhance class {@link org.apache.dubbo.monitor.support.MonitorFilter#invoke(Invoker,
+ * Invocation)}. the trace context transport to the provider side by {@link RpcContext#getServerAttachment}.but all the version
+ * of dubbo framework below 3.0.0 don't support {@link RpcContext#getClientAttachment}, we support another way to support it.
+ */
+public class DubboInterceptor implements InstanceMethodsAroundInterceptor {
+
+    public static final String ARGUMENTS = "arguments";
+
+    /**
+     * <h2>Consumer:</h2> The serialized trace context data will
+     * inject to the {@link RpcContext#getClientAttachment} for transport to provider side.
+     * <p>
+     * <h2>Provider:</h2> The serialized trace context data will extract from
+     * {@link RpcContext#getServerAttachment}. current trace segment will ref if the serialize context data is not null.
+     */
+    @Override
+    public void beforeMethod(EnhancedInstance objInst, Method method, Object[] allArguments, Class<?>[] argumentsTypes,
+                             MethodInterceptResult result) throws Throwable {
+        Invoker invoker = (Invoker) allArguments[0];
+        Invocation invocation = (Invocation) allArguments[1];
+
+        RpcServiceContext serviceContext = RpcContext.getServiceContext();
+        boolean isConsumer;
+        URL url = serviceContext.getUrl();
+        if (url == null) {
+            url = serviceContext.getConsumerUrl();
+            isConsumer = url != null;
+        } else {
+            isConsumer = serviceContext.isConsumerSide();
+        }
```

Comment:
Should not `serviceContext.isConsumerSide()` have high priority? I can see from the existing plugin, `url == null` is not an issue, it is just for server side operation name.

Reply:
Is there anything different from Dubbo 2.7? I can see this in the existing codes.

```java
boolean isConsumer = rpcContext.isConsumerSide();
```

Reply:
https://github.com/apache/dubbo/blob/5a5e3f3fd7d0482b6124bd5b52b10f4900856438/dubbo-rpc/dubbo-rpc-api/src/main/java/org/apache/dubbo/rpc/RpcServiceContext.java#L142-L145

```java
@Override
public boolean isConsumerSide() {
    return getUrl().getSide(PROVIDER_SIDE).equals(CONSUMER_SIDE);
}

@Override
public boolean isProviderSide() {
    return !isConsumerSide();
}
```

`isConsumerSide` depends on `getUrl`, when I debugging locally, the url sometime is `null`, caused NullPointerException. So I think apm should not crash rpc, and add this checking.

Reply:
Interesting, I think you are better to report a bug to Dubbo. I am OK for your plugin codes, but Dubbo users may complain, and `sometimes null` seems a bug :)

Reply:
Then please fix the format(unexpected code changes), I am good.